### PR TITLE
feat(utapi): Add listFiles to UTAPI

### PR DESCRIPTION
**Pull request by markflorkowski**

### CODE REVIEW

File: packages/uploadthing/src/sdk/index.ts [hunk 1]
```diff
@@ -55,8 +60,40 @@ export const getFileUrls = async (fileKeys: string[] | string) => {
...
+    },
+  });
+
+  if (!res.ok) {
+    throw new Error("Failed to list files");
+  }
+
+  return res
+    .json()
+    .then(({ data }) => data as { fileKeys: string[]; fileIds: string[] });
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/index.ts:98]: Should the return type not be `{ key: string, id: string }[]`

Comment (markflorkowski) [reply packages/uploadthing/src/sdk/index.ts:98]: Currently that endpoint is actually returning

```
{
  fileIds: {
     fileIds: string[]
     fileKeys: string[]
  }
}
```

🤦 It should probably return something more reasonable like you describe though....

Comment (markflorkowski) [reply packages/uploadthing/src/sdk/index.ts:98]: image

File: docs/src/pages/api-reference/server.mdx [hunk 2]
```diff
@@ -188,3 +188,25 @@ const multipleUrls = await utapi.getFileUrls([
 | Arguments | Type                 | Required | Description                                      |
 | :-------- | :------------------- | :------- | :----------------------------------------------- |
 | input     | string \|\| string[] | Yes      | The fileKey or fileKeys you want to get URLs for |
+
+### `listFiles`
+
+`listFiles` takes no input and returns an array of objects containing file ids
+and keys for all files that have bee uploaded to the application your API key
```
Comment (juliusmarminge) [inline docs/src/pages/api-reference/server.mdx:195]:
```suggestion
and keys for all files that have been uploaded to the application your API key
```

File: packages/uploadthing/src/sdk/index.ts [hunk 3]
```diff
@@ -55,8 +60,39 @@ export const getFileUrls = async (fileKeys: string[] | string) => {
...
+
   return res.json().then(({ data }) => data as { key: string; url: string }[]);
 };
+
+/**
+ * Request file list from UploadThing storage.
+ *
+ * @example
+ * const data = await listFiles();
+ * console.log(data); // {fileKeys: ["2e0fdb64-9957-4262-8e45-f372ba903ac8_image.jpg"], fileIds: ["2e0fdb64-9957-4262-8e45-f372ba903ac8"]}
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/index.ts:76]:
```suggestion
 * console.log(data); // { key: "2e0fdb64-9957-4262-8e45-f372ba903ac8_image.jpg", id: "2e0fdb64-9957-4262-8e45-f372ba903ac8" }
```

File: packages/uploadthing/src/sdk/index.ts [hunk 4]
```diff
@@ -55,8 +60,39 @@ export const getFileUrls = async (fileKeys: string[] | string) => {
...
+      "x-uploadthing-api-key": UT_SECRET,
+      "x-uploadthing-version": UPLOADTHING_VERSION,
+    },
+  });
+
+  if (!res.ok) {
+    throw new Error("Failed to list files");
+  }
+
+  return res.json().then(({ data: { files } }) => {
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/index.ts:95]: This doesn't match the response from the infra

```suggestion
  return res.json().then(({ files }) => {
```

Comment (juliusmarminge) [reply packages/uploadthing/src/sdk/index.ts:95]: CleanShot 2023-07-17 at 23 40 23@2x

File: docs/src/pages/api-reference/server.mdx [hunk 5]
```diff
@@ -188,3 +188,25 @@ const multipleUrls = await utapi.getFileUrls([
...
+### `listFiles`
+
+`listFiles` takes no input and returns an array of objects containing file ids
+and keys for all files that have been uploaded to the application your API key
+corresponds to.
+
+```ts
+import { utapi } from "uploadthing/server";
+
+const files = utapi.listFiles();
```
Comment (juliusmarminge) [inline docs/src/pages/api-reference/server.mdx:201]:
```suggestion
const files = await utapi.listFiles();
```

File: packages/uploadthing/src/sdk/index.ts [hunk 6]
```diff
@@ -55,8 +60,39 @@ export const getFileUrls = async (fileKeys: string[] | string) => {
...
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+      "x-uploadthing-api-key": UT_SECRET,
+      "x-uploadthing-version": UPLOADTHING_VERSION,
+    },
+  });
+
+  if (!res.ok) {
+    throw new Error("Failed to list files");
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/index.ts:92]: this error should probably be improved.

what i did for the server uploads was the following:

```ts
  const json = (await res.json()) as
    | { data: { key: string; url: string }[] }
    | { error: string };

  if (!res.ok || "error" in json) {
    const message = "error" in json ? json.error : "Unknown error";
    throw new Error(message);
  }

  return json.data;
```

Comment (juliusmarminge) [reply packages/uploadthing/src/sdk/index.ts:92]: would obviously require a follow up PR to the infra

Comment (markflorkowski) [reply packages/uploadthing/src/sdk/index.ts:92]: Can do!

File: packages/uploadthing/src/sdk/index.ts [hunk 7]
```diff
@@ -55,8 +60,42 @@ export const getFileUrls = async (fileKeys: string[] | string) => {
...
+  const json = (await res.json()) as
+    | { files: { key: string; id: string }[] }
+    | { error: string };
+
+  if (!res.ok || "error" in json) {
+    const message = "error" in json ? json.error : "Unknown error";
+    throw new Error(message);
+  }
+
+  return json.files as { id: string; key: string }[];
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/index.ts:100]: Unnecessary cast

File: packages/uploadthing/src/sdk/index.ts [hunk 8]
```diff
@@ -55,8 +60,42 @@ export const getFileUrls = async (fileKeys: string[] | string) => {
...
+    headers: {
+      "Content-Type": "application/json",
+      "x-uploadthing-api-key": UT_SECRET,
+      "x-uploadthing-version": UPLOADTHING_VERSION,
+    },
+  });
+
+  const json = (await res.json()) as
+    | { files: { key: string; id: string }[] }
+    | { error: string };
```
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/index.ts:93]: I saw that you also returned a `message` prop on the infra side. Which is more descriptive to use?

Comment (markflorkowski) [reply packages/uploadthing/src/sdk/index.ts:93]: Honestly not sure what will come out of the message, could be more descriptive, could be unintelligible garbage. 😅

File: packages/uploadthing/src/sdk/index.ts (hunk 7 above)
Comment (juliusmarminge) [inline packages/uploadthing/src/sdk/index.ts:100]:
```suggestion
  return json.files;
```

File: packages/uploadthing/src/sdk/index.ts (hunk 8 above)
Comment (juliusmarminge) [reply packages/uploadthing/src/sdk/index.ts:93]: aight let's just go with this then

**Review by juliusmarminge — APPROVED** [review]

**Pull request merged**